File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/ModelFallbackGenerator.java
package com.solutions.alphil.zambiajobalerts.classes;

import android.text.TextUtils;

import com.solutions.alphil.zambiajobalerts.classes.ApiConfig;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class ModelFallbackGenerator {

    public interface GenerationListener {
        void onSuccess(String generatedText);
        void onFailure(String message);
    }

    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    private final OkHttpClient client = new OkHttpClient();

    public void generateText(String prompt, GenerationListener listener) {
        List<ModelProviderConfig> providers = buildProviders();
        tryProvider(0, providers, prompt, listener);
    }

    private void tryProvider(int index, List<ModelProviderConfig> providers, String prompt,
                             GenerationListener listener) {
        if (index >= providers.size()) {
            fallbackToPollinations(prompt, listener);
            return;
        }

        ModelProviderConfig config = providers.get(index);
        RequestBody body;

        try {
            JSONObject payload = new JSONObject();
            payload.put("model", config.model);

            JSONObject systemMessage = new JSONObject();
            systemMessage.put("role", "system");
            systemMessage.put("content", "You are a professional CV and cover letter writing assistant.");

            JSONObject userMessage = new JSONObject();
            userMessage.put("role", "user");
            userMessage.put("content", prompt);

            JSONArray messages = new JSONArray();
            messages.put(systemMessage);
            messages.put(userMessage);
            payload.put("messages", messages);

            body = RequestBody.create(payload.toString(), JSON);
        } catch (JSONException e) {
            listener.onFailure("Failed to build request: " + e.getMessage());
            return;
        }

        Request.Builder requestBuilder = new Request.Builder().url(config.endpoint).post(body);
        if (!TextUtils.isEmpty(config.apiKey)) {
            requestBuilder.header("Authorization", "Bearer " + config.apiKey);
        }

        Request request = requestBuilder.build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                String next = "Provider " + config.name + " request failed. Trying next model.";
                tryProvider(index + 1, providers, prompt, new DelegateListener(listener, next));
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                String bodyString = response.body() != null ? response.body().string() : "";
                if (!response.isSuccessful()) {
                    String next = "Provider " + config.name + " returned " + response.code() + ". Trying next model.";
                    tryProvider(index + 1, providers, prompt, new DelegateListener(listener, next));
                    return;
                }

                String text = parseOpenAIStyleResponse(bodyString);
                if (text == null || text.trim().isEmpty()) {
                    String next = "Provider " + config.name + " returned empty output. Trying next model.";
                    tryProvider(index + 1, providers, prompt, new DelegateListener(listener, next));
                    return;
                }

                listener.onSuccess(text.trim());
            }
        });
    }

    private void fallbackToPollinations(String prompt, GenerationListener listener) {
        List<String> pollinationModels = splitCsvValues(ApiConfig.TEXT_POLLINATIONS_MODELS);
        if (pollinationModels.isEmpty()) {
            pollinationModels.add("openai");
        }

        tryPollinations(prompt, 0, pollinationModels, listener);
    }

    private void tryPollinations(String prompt, int index, List<String> models,
                                GenerationListener listener) {
        if (index >= models.size()) {
            listener.onFailure("Pollinations fallback failed for all available models.");
            return;
        }

        String model = models.get(index);
        String base = ApiConfig.TEXT_POLLINATIONS_ENDPOINT;
        String encodedPrompt = encodePrompt(prompt);
        String url = model == null || model.trim().isEmpty()
                ? base + encodedPrompt
                : base + model.trim() + "/" + encodedPrompt;

        Request request = new Request.Builder().url(url).get().build();
        final String modelName = model == null || model.trim().isEmpty() ? "default" : model.trim();
        final int nextIndex = index + 1;
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                String next = "Pollinations model " + modelName + " request failed. Trying next model.";
                tryPollinations(prompt, nextIndex, models, new DelegateListener(listener, next));
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                String bodyString = response.body() != null ? response.body().string() : "";
                if (!response.isSuccessful()) {
                    String next = "Pollinations model " + modelName + " returned " + response.code() + ". Trying next model.";
                    tryPollinations(prompt, nextIndex, models, new DelegateListener(listener, next));
                    return;
                }

                if (bodyString == null || bodyString.trim().isEmpty()) {
                    String next = "Pollinations model " + modelName + " returned empty output. Trying next model.";
                    tryPollinations(prompt, nextIndex, models, new DelegateListener(listener, next));
                    return;
                }

                if (isPollinationsErrorPayload(bodyString)) {
                    String next = "Pollinations model " + modelName + " returned error payload. Trying next model.";
                    tryPollinations(prompt, nextIndex, models, new DelegateListener(listener, next));
                    return;
                }

                if (isUnusablePollinationsPayload(bodyString)) {
                    String next = "Pollinations model " + modelName + " returned non-text response. Trying next model.";
                    tryPollinations(prompt, nextIndex, models, new DelegateListener(listener, next));
                    return;
                }

                listener.onSuccess(bodyString.trim());
            }
        });
    }

    private boolean isPollinationsErrorPayload(String raw) {
        String trimmed = raw == null ? "" : raw.trim();
        if (trimmed.startsWith("{")) {
            try {
                JSONObject json = new JSONObject(trimmed);
                return json.has("error");
            } catch (JSONException ignored) {
                return false;
            }
        }
        if (trimmed.contains("Queue full for IP") || trimmed.contains("\"status\":429")) {
            return true;
        }
        return false;
    }

    private boolean isUnusablePollinationsPayload(String raw) {
        String trimmed = raw == null ? "" : raw.trim().toLowerCase(java.util.Locale.ROOT);
        return trimmed.startsWith("<!doctype html>") || trimmed.startsWith("<html");
    }

    private String parseOpenAIStyleResponse(String raw) {
        try {
            JSONObject json = new JSONObject(raw);
            JSONArray choices = json.optJSONArray("choices");
            if (choices == null || choices.length() == 0) {
                return null;
            }

            JSONObject choice = choices.optJSONObject(0);
            if (choice == null) {
                return null;
            }

            JSONObject message = choice.optJSONObject("message");
            if (message != null) {
                return message.optString("content", "");
            }

            return choice.optString("text", "");
        } catch (JSONException ignored) {
            return null;
        }
    }

    private List<ModelProviderConfig> buildProviders() {
        List<ModelProviderConfig> providers = new ArrayList<>();

        addProvidersIfAvailable(providers,
                ApiConfig.PRIMARY_TEXT_MODEL_URL,
                ApiConfig.PRIMARY_TEXT_MODEL_KEY,
                ApiConfig.PRIMARY_TEXT_MODEL_NAME,
                "Primary");

        addProvidersIfAvailable(providers,
                ApiConfig.SECONDARY_TEXT_MODEL_URL,
                ApiConfig.SECONDARY_TEXT_MODEL_KEY,
                ApiConfig.SECONDARY_TEXT_MODEL_NAME,
                "Secondary");

        return providers;
    }

    private void addProvidersIfAvailable(List<ModelProviderConfig> providers,
                                        String endpoints,
                                        String apiKeys,
                                        String models,
                                        String label) {
        List<String> endpointList = splitCsvValues(endpoints);
        if (endpointList.isEmpty()) {
            return;
        }

        List<String> keyList = splitCsvValues(apiKeys);
        List<String> modelList = splitCsvValues(models);
        int providerCount = Math.max(endpointList.size(),
                Math.max(keyList.size(), modelList.size()));

        for (int i = 0; i < providerCount; i++) {
            String endpoint = pickValue(endpointList, i, endpointList.get(endpointList.size() - 1));
            if (endpoint.isEmpty()) {
                continue;
            }
            String apiKey = pickValue(keyList, i, "");
            String model = pickValue(modelList, i, "gpt-3.5-turbo");
            String providerLabel = endpointList.size() > 1 || keyList.size() > 1 || modelList.size() > 1
                    ? label + " #" + (i + 1)
                    : label;
            providers.add(new ModelProviderConfig(endpoint, apiKey, model, providerLabel));
        }
    }

    private List<String> splitCsvValues(String raw) {
        List<String> values = new ArrayList<>();
        if (raw == null || raw.trim().isEmpty()) {
            return values;
        }

        String[] tokens = raw.trim().split("\\s*[,;\\n]\\s*");
        for (String token : tokens) {
            if (token == null || token.trim().isEmpty()) {
                continue;
            }
            values.add(token.trim());
        }
        return values;
    }

    private String pickValue(List<String> values, int index, String fallback) {
        if (values == null || values.isEmpty()) {
            return fallback;
        }

        if (values.size() == 1) {
            return values.get(0);
        }

        if (index < values.size()) {
            return values.get(index);
        }

        return fallback;
    }

    private String encodePrompt(String prompt) {
        try {
            return URLEncoder.encode(prompt, StandardCharsets.UTF_8.toString()).replace("+", "%20");
        } catch (UnsupportedEncodingException e) {
            return "";
        }
    }

    private static class ModelProviderConfig {
        private final String endpoint;
        private final String apiKey;
        private final String model;
        private final String name;

        private ModelProviderConfig(String endpoint, String apiKey, String model, String name) {
            this.endpoint = endpoint;
            this.apiKey = apiKey;
            this.model = model;
            this.name = name;
        }
    }

    private static class DelegateListener implements GenerationListener {
        private final GenerationListener target;
        private final String nextMessage;

        private DelegateListener(GenerationListener target, String nextMessage) {
            this.target = target;
            this.nextMessage = nextMessage;
        }

        @Override
        public void onSuccess(String generatedText) {
            target.onSuccess(generatedText);
        }

        @Override
        public void onFailure(String message) {
            target.onFailure(nextMessage + " " + message);
        }
    }
}

File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/ModelFallbackGenerator.kt
package com.solutions.alphil.zambiajobalerts.classes

import android.text.TextUtils
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Locale
import kotlin.math.max

class ModelFallbackGenerator {
    private val client = OkHttpClient()

    fun generateText(prompt: String, listener: GenerationListener) {
        val providers = buildProviders()
        tryProvider(0, providers, prompt, listener)
    }

    private fun tryProvider(
        index: Int,
        providers: List<ModelProviderConfig>,
        prompt: String,
        listener: GenerationListener,
    ) {
        if (index >= providers.size) {
            fallbackToPollinations(prompt, listener)
            return
        }

        val config = providers[index]
        val body = try {
            val payload = JSONObject()
            payload.put("model", config.model)

            val systemMessage = JSONObject()
            systemMessage.put("role", "system")
            systemMessage.put("content", "You are a professional CV and cover letter writing assistant.")

            val userMessage = JSONObject()
            userMessage.put("role", "user")
            userMessage.put("content", prompt)

            val messages = JSONArray()
            messages.put(systemMessage)
            messages.put(userMessage)
            payload.put("messages", messages)

            payload.toString().toRequestBody(JSON)
        } catch (error: JSONException) {
            listener.onFailure("Failed to build request: ${error.message}")
            return
        }

        val requestBuilder = Request.Builder().url(config.endpoint).post(body)
        if (!TextUtils.isEmpty(config.apiKey)) {
            requestBuilder.header("Authorization", "Bearer ${config.apiKey}")
        }

        val request = requestBuilder.build()
        client.newCall(request).enqueue(
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    val next = "Provider ${config.name} request failed. Trying next model."
                    tryProvider(index + 1, providers, prompt, DelegateListener(listener, next))
                }

                override fun onResponse(call: Call, response: Response) {
                    response.use {
                        val bodyString = it.body.string()
                        if (!it.isSuccessful) {
                            val next = "Provider ${config.name} returned ${it.code}. Trying next model."
                            tryProvider(index + 1, providers, prompt, DelegateListener(listener, next))
                            return
                        }

                        val text = parseOpenAIStyleResponse(bodyString)
                        if (text.isNullOrBlank()) {
                            val next = "Provider ${config.name} returned empty output. Trying next model."
                            tryProvider(index + 1, providers, prompt, DelegateListener(listener, next))
                            return
                        }

                        listener.onSuccess(text.trim())
                    }
                }
            },
        )
    }

    private fun fallbackToPollinations(prompt: String, listener: GenerationListener) {
        val pollinationModels = splitCsvValues(ApiConfig.TEXT_POLLINATIONS_MODELS).toMutableList()
        if (pollinationModels.isEmpty()) {
            pollinationModels.add("openai")
        }

        tryPollinations(prompt, 0, pollinationModels, listener)
    }

    private fun tryPollinations(
        prompt: String,
        index: Int,
        models: List<String>,
        listener: GenerationListener,
    ) {
        if (index >= models.size) {
            listener.onSuccess(LocalDocumentTemplateGenerator.generate(prompt))
            return
        }

        val model = models[index]
        val base = ApiConfig.TEXT_POLLINATIONS_ENDPOINT
        val encodedPrompt = encodePrompt(prompt)
        val url = if (model.trim().isEmpty()) {
            base + encodedPrompt
        } else {
            base + model.trim() + "/" + encodedPrompt
        }

        val request = Request.Builder().url(url).get().build()
        val modelName = if (model.trim().isEmpty()) "default" else model.trim()
        val nextIndex = index + 1
        client.newCall(request).enqueue(
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    val next = "Pollinations model $modelName request failed. Trying next model."
                    tryPollinations(prompt, nextIndex, models, DelegateListener(listener, next))
                }

                override fun onResponse(call: Call, response: Response) {
                    response.use {
                        val bodyString = it.body.string()
                        if (!it.isSuccessful) {
                            val next = "Pollinations model $modelName returned ${it.code}. Trying next model."
                            tryPollinations(prompt, nextIndex, models, DelegateListener(listener, next))
                            return
                        }

                        if (bodyString.isBlank()) {
                            val next = "Pollinations model $modelName returned empty output. Trying next model."
                            tryPollinations(prompt, nextIndex, models, DelegateListener(listener, next))
                            return
                        }

                        if (isPollinationsErrorPayload(bodyString)) {
                            val next = "Pollinations model $modelName returned error payload. Trying next model."
                            tryPollinations(prompt, nextIndex, models, DelegateListener(listener, next))
                            return
                        }

                        if (isUnusablePollinationsPayload(bodyString)) {
                            val next = "Pollinations model $modelName returned non-text response. Trying next model."
                            tryPollinations(prompt, nextIndex, models, DelegateListener(listener, next))
                            return
                        }

                        listener.onSuccess(bodyString.trim())
                    }
                }
            },
        )
    }

    private fun isPollinationsErrorPayload(raw: String?): Boolean {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.startsWith("{")) {
            try {
                val json = JSONObject(trimmed)
                return json.has("error")
            } catch (ignored: JSONException) {
                return false
            }
        }
        return trimmed.contains("Queue full for IP") || trimmed.contains("\"status\":429")
    }

    private fun isUnusablePollinationsPayload(raw: String?): Boolean {
        val trimmed = raw?.trim().orEmpty().lowercase(Locale.ROOT)
        return trimmed.startsWith("<!doctype html>") || trimmed.startsWith("<html")
    }

    private fun parseOpenAIStyleResponse(raw: String): String? =
        try {
            val json = JSONObject(raw)
            val choices = json.optJSONArray("choices")
            if (choices == null || choices.length() == 0) {
                null
            } else {
                val choice = choices.optJSONObject(0)
                val message = choice?.optJSONObject("message")
                message?.optString("content", "") ?: choice?.optString("text", "")
            }
        } catch (ignored: JSONException) {
            null
        }

    private fun buildProviders(): List<ModelProviderConfig> {
        val providers = mutableListOf<ModelProviderConfig>()

        addProvidersIfAvailable(
            providers,
            ApiConfig.PRIMARY_TEXT_MODEL_URL,
            ApiConfig.PRIMARY_TEXT_MODEL_KEY,
            ApiConfig.PRIMARY_TEXT_MODEL_NAME,
            "Primary",
        )

        addProvidersIfAvailable(
            providers,
            ApiConfig.SECONDARY_TEXT_MODEL_URL,
            ApiConfig.SECONDARY_TEXT_MODEL_KEY,
            ApiConfig.SECONDARY_TEXT_MODEL_NAME,
            "Secondary",
        )

        return providers
    }

    private fun addProvidersIfAvailable(
        providers: MutableList<ModelProviderConfig>,
        endpoints: String?,
        apiKeys: String?,
        models: String?,
        label: String,
    ) {
        val endpointList = splitCsvValues(endpoints)
        if (endpointList.isEmpty()) {
            return
        }

        val keyList = splitCsvValues(apiKeys)
        val modelList = splitCsvValues(models)
        val providerCount = max(endpointList.size, max(keyList.size, modelList.size))

        for (index in 0 until providerCount) {
            val endpoint = pickValue(endpointList, index, endpointList[endpointList.size - 1])
            if (endpoint.isEmpty()) {
                continue
            }
            val apiKey = pickValue(keyList, index, "")
            val model = pickValue(modelList, index, "gpt-3.5-turbo")
            val providerLabel = if (endpointList.size > 1 || keyList.size > 1 || modelList.size > 1) {
                "$label #${index + 1}"
            } else {
                label
            }
            providers.add(ModelProviderConfig(endpoint, apiKey, model, providerLabel))
        }
    }

    private fun splitCsvValues(raw: String?): List<String> {
        if (raw.isNullOrBlank()) {
            return emptyList()
        }

        val values = mutableListOf<String>()
        raw.trim().split(Regex("\\s*[,;\\n]\\s*")).forEach { token ->
            if (token.isNotBlank()) {
                values.add(token.trim())
            }
        }
        return values
    }

    private fun pickValue(values: List<String>?, index: Int, fallback: String): String {
        if (values.isNullOrEmpty()) {
            return fallback
        }
        if (values.size == 1) {
            return values[0]
        }
        if (index < values.size) {
            return values[index]
        }
        return fallback
    }

    private fun encodePrompt(prompt: String): String =
        try {
            URLEncoder.encode(prompt, StandardCharsets.UTF_8.toString()).replace("+", "%20")
        } catch (error: Exception) {
            ""
        }

    fun interface GenerationListener {
        fun onSuccess(generatedText: String)
        fun onFailure(message: String) = Unit
    }

    private data class ModelProviderConfig(
        val endpoint: String,
        val apiKey: String,
        val model: String,
        val name: String,
    )

    private class DelegateListener(
        private val target: GenerationListener,
        private val nextMessage: String,
    ) : GenerationListener {
        override fun onSuccess(generatedText: String) {
            target.onSuccess(generatedText)
        }

        override fun onFailure(message: String) {
            target.onFailure("$nextMessage $message")
        }
    }

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaTypeOrNull()
    }
}

private object LocalDocumentTemplateGenerator {
    fun generate(prompt: String): String {
        val fields = parseFields(prompt)
        return if (prompt.lowercase(Locale.ROOT).contains("cover letter")) {
            buildCoverLetter(fields)
        } else {
            buildCv(fields)
        }
    }

    private fun parseFields(prompt: String): Map<String, String> {
        val fields = linkedMapOf<String, String>()
        prompt.lines().forEach { line ->
            val separator = line.indexOf(':')
            if (separator > 0 && separator < line.length - 1) {
                val key = line.substring(0, separator).trim().lowercase(Locale.ROOT)
                val value = clean(line.substring(separator + 1))
                if (value.isNotEmpty()) {
                    fields[key] = value
                }
            }
        }
        return fields
    }

    private fun buildCv(fields: Map<String, String>): String {
        val name = field(fields, "applicant name").ifBlank { "Applicant" }
        val email = field(fields, "email")
        val phone = field(fields, "phone")
        val role = field(fields, "write a modern cv/resume for the role", "job title", "role")
            .ifBlank { "the target role" }
        val education = field(fields, "education")
        val experience = field(fields, "work experience", "relevant experience")
        val skills = splitList(field(fields, "skills"))
        val notes = field(fields, "extra notes")
        val jobSummary = field(fields, "job summary")
        val jobLocation = field(fields, "job location")
        val contactLine = listOf(email, phone).filter { it.isNotBlank() }.joinToString(" | ")

        return buildString {
            appendLine(name)
            if (contactLine.isNotBlank()) {
                appendLine(contactLine)
            }
            appendLine()
            appendLine("Professional Summary")
            appendLine(
                "Interview-ready candidate targeting $role with a practical background aligned to the needs of the position.",
            )
            when {
                experience.isNotBlank() -> appendLine("Experience focus: $experience")
                jobSummary.isNotBlank() -> appendLine("Role focus: $jobSummary")
            }
            if (jobLocation.isNotBlank()) {
                appendLine("Preferred location: $jobLocation")
            }
            appendLine()
            appendLine("Key Skills")
            if (skills.isEmpty()) {
                appendLine("- Clear communication and professional conduct")
                appendLine("- Problem solving and reliable task execution")
                appendLine("- Team collaboration and customer-focused service")
            } else {
                skills.forEach { appendLine("- $it") }
            }
            appendLine()
            appendLine("Work Experience")
            appendLine(
                if (experience.isBlank()) {
                    "Relevant experience can be tailored further with previous employer names, dates, and measurable achievements."
                } else {
                    experience
                },
            )
            appendLine()
            appendLine("Education")
            appendLine(if (education.isBlank()) "Education details available on request." else education)
            if (notes.isNotBlank()) {
                appendLine()
                appendLine("Additional Information")
                appendLine(notes)
            }
            appendLine()
            appendLine("References")
            appendLine("Available on request.")
        }.trim()
    }

    private fun buildCoverLetter(fields: Map<String, String>): String {
        val name = field(fields, "applicant name").ifBlank { "Applicant" }
        val email = field(fields, "email")
        val phone = field(fields, "phone")
        val company = field(fields, "company", "job company").ifBlank { "your organisation" }
        val role = field(fields, "role", "job title", "write a modern cv/resume for the role")
            .ifBlank { "the advertised position" }
        val experience = field(fields, "relevant experience", "work experience")
        val notes = field(fields, "extra notes")
        val jobSummary = field(fields, "job summary")
        val tone = field(fields, "tone").replace('_', ' ').ifBlank { "professional" }
        val contactLine = listOf(email, phone).filter { it.isNotBlank() }.joinToString(" | ")

        return buildString {
            appendLine(name)
            if (contactLine.isNotBlank()) {
                appendLine(contactLine)
            }
            appendLine()
            appendLine("Dear Hiring Manager,")
            appendLine()
            appendLine(
                "I am writing to apply for the $role position at $company. I am interested in this opportunity because it matches my skills, work ethic, and commitment to delivering dependable results.",
            )
            appendLine()
            if (experience.isNotBlank()) {
                appendLine(
                    "My relevant experience includes $experience. This background has prepared me to contribute quickly, communicate clearly, and handle responsibilities with care.",
                )
            } else {
                appendLine(
                    "I bring a strong willingness to learn, a professional approach to work, and the discipline needed to contribute effectively in this role.",
                )
            }
            if (jobSummary.isNotBlank()) {
                appendLine()
                appendLine("Based on the job details, I understand that the role requires focus in areas such as $jobSummary.")
            }
            if (notes.isNotBlank()) {
                appendLine()
                appendLine(notes)
            }
            appendLine()
            appendLine(
                "I would welcome the opportunity to discuss how my background and $tone approach can support $company. Thank you for considering my application.",
            )
            appendLine()
            appendLine("Sincerely,")
            appendLine(name)
        }.trim()
    }

    private fun field(fields: Map<String, String>, vararg keys: String): String {
        keys.forEach { key ->
            val value = fields[key]?.trim()
            if (!value.isNullOrEmpty()) {
                return clean(value)
            }
        }
        return ""
    }

    private fun splitList(raw: String): List<String> =
        raw.split(Regex("\\s*[,;\\n]\\s*"))
            .map { clean(it) }
            .filter { it.isNotBlank() }

    private fun clean(value: String): String =
        value
            .replace(Regex("<[^>]*>"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()
}
